File: src/schemas/task.ts
import { z } from 'zod';
import { BoxInputSchema, PointInputSchema } from '@/schemas/annotation';
import { OpenVolumeResponseSchema } from '@/schemas/volume';

export const TaskKindSchema = z.enum(['Segment', 'Detect']);
export type TaskKind = z.infer<typeof TaskKindSchema>;

export const TaskStatusSchema = z.enum(['Pending', 'Running', 'Done', 'Failed', 'Cancelled']);
export type TaskStatus = z.infer<typeof TaskStatusSchema>;

export const DetectTargetSchema = z.object({
  label_id: z.string(),
  sub_labels: z.array(z.string()),
});
export type DetectTarget = z.infer<typeof DetectTargetSchema>;

export const TaskParamsSchema = z.object({
  use_prev_mask: z.boolean(),
  multimask_output: z.boolean(),
  targets: z.array(DetectTargetSchema),
});
export type TaskParams = z.infer<typeof TaskParamsSchema>;

export const TaskSchema = z.object({
  id: z.string(),
  name: z.string(),
  kind: TaskKindSchema,
  status: TaskStatusSchema,
  enabled: z.boolean(),
  order: z.number(),
  volume_id: z.string(),
  range_start: z.number(),
  range_end: z.number(),
  params: TaskParamsSchema,
  error: z.string().nullable(),
  progress_current: z.number(),
  progress_total: z.number(),
  created_at: z.number(),
  updated_at: z.number(),
});
export type Task = z.infer<typeof TaskSchema>;

export const TaskCreateInputSchema = z.object({
  name: z.string().min(1),
  kind: TaskKindSchema,
  volumeId: z.string(),
  rangeStart: z.number().int().nonnegative(),
  rangeEnd: z.number().int().nonnegative(),
  params: TaskParamsSchema,
});
export type TaskCreateInput = z.infer<typeof TaskCreateInputSchema>;

export const TaskPatchSchema = z.object({
  name: z.string().optional(),
  enabled: z.boolean().optional(),
  order: z.number().optional(),
  rangeStart: z.number().optional(),
  rangeEnd: z.number().optional(),
  params: TaskParamsSchema.optional(),
});
export type TaskPatch = z.infer<typeof TaskPatchSchema>;

export const TaskOrderInputSchema = z.object({
  id: z.string(),
  order: z.number(),
});
export type TaskOrderInput = z.infer<typeof TaskOrderInputSchema>;

export const AiObjectInputSchema = z.object({
  id: z.string(),
  label_id: z.string(),
  points: z.array(PointInputSchema),
  boxes: z.array(BoxInputSchema),
});
export type AiObjectInput = z.infer<typeof AiObjectInputSchema>;

export const AiRecognizeResponseSchema = z.object({
  maskHash: z.string(),
});
export type AiRecognizeResponse = z.infer<typeof AiRecognizeResponseSchema>;

export { OpenVolumeResponseSchema as CurrentVolumeSchema };
export type CurrentVolume = z.infer<typeof OpenVolumeResponseSchema>;

File: crates/db/src/models/task.rs
/// 自动打标任务类型。
#[derive(Debug, Clone, PartialEq, toasty::Embed, serde::Serialize, serde::Deserialize)]
pub enum TaskKind {
    #[column(variant = "segment")]
    Segment,
    #[column(variant = "detect")]
    Detect,
}

/// 自动打标任务状态。
#[derive(Debug, Clone, PartialEq, toasty::Embed, serde::Serialize, serde::Deserialize)]
pub enum TaskStatus {
    #[column(variant = "pending")]
    Pending,
    #[column(variant = "running")]
    Running,
    #[column(variant = "done")]
    Done,
    #[column(variant = "failed")]
    Failed,
    #[column(variant = "cancelled")]
    Cancelled,
}

/// 任务执行参数（序列化进 `Task.params` JSON 列）。
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct TaskParams {
    pub use_prev_mask: bool,
    pub targets: Vec<DetectTarget>,
}

/// detect 任务的单个目标标签。
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct DetectTarget {
    pub label_id: uuid::Uuid,
    /// 空表示使用 `label.name` 作为 category。
    pub sub_labels: Vec<String>,
}

#[derive(Debug, toasty::Model, serde::Serialize)]
pub struct Task {
    #[key]
    #[auto]
    pub id: uuid::Uuid,

    pub name: String,

    pub kind: TaskKind,

    pub status: TaskStatus,

    pub enabled: bool,

    pub order: i32,

    pub volume_id: String,

    pub range_start: i32,

    pub range_end: i32,

    #[column(type = "text")]
    pub params: toasty::Json<TaskParams>,

    pub error: Option<String>,

    pub progress_current: i32,

    pub progress_total: i32,

    pub created_at: i64,

    pub updated_at: i64,
}
